x-pack/plugins/actions/server/plugin.ts

```diff
@@ -0,0 +1,242 @@
+    });
+
+    plugins.event_log.registerProviderActions(EVENT_LOG_PROVIDER, Object.values(EVENT_LOG_ACTIONS));
+    this.eventLogger = plugins.event_log.getLogger({
+      event: { provider: EVENT_LOG_PROVIDER },
+    });
+
+    const actionExecutor = new ActionExecutor();
+    const taskRunnerFactory = new TaskRunnerFactory(actionExecutor);
+    const actionsConfigUtils = getActionsConfigurationUtilities(
+      (await this.config) as ActionsConfig
+    );
+    const actionTypeRegistry = new ActionTypeRegistry({
+      taskRunnerFactory,
+      taskManager: plugins.taskManager,
+      actionsConfigUtils,
+    });
+    this.taskRunnerFactory = taskRunnerFactory;
+    this.actionTypeRegistry = actionTypeRegistry;
+    this.serverBasePath = core.http.basePath.serverBasePath;
+    this.actionExecutor = actionExecutor;
+    this.adminClient = core.elasticsearch.adminClient;
+    this.spaces = plugins.spaces?.spacesService;
+
+    registerBuiltInActionTypes({
+      logger: this.logger,
+      actionTypeRegistry,
+      actionsConfigUtils,
+    });
+
+    core.http.registerRouteHandlerContext(
+      'actions',
+      this.createRouteHandlerContext(await this.kibanaIndex)
+    );
+
+    // Routes
+    this.licenseState = new LicenseState(plugins.licensing.license$);
+    const router = core.http.createRouter();
+    createActionRoute(router, this.licenseState);
+    deleteActionRoute(router, this.licenseState);
+    getActionRoute(router, this.licenseState);
+    findActionRoute(router, this.licenseState);
+    updateActionRoute(router, this.licenseState);
+    listActionTypesRoute(router, this.licenseState);
+    executeActionRoute(router, this.licenseState, actionExecutor);
+
+    return {
+      registerType: actionTypeRegistry.register.bind(actionTypeRegistry),
+    };
+  }
+
+  public start(core: CoreStart, plugins: ActionsPluginsStart): PluginStartContract {
+    const { logger, actionExecutor, actionTypeRegistry, taskRunnerFactory } = this;
+
+    actionExecutor!.initialize({
+      logger,
+      eventLogger: this.eventLogger!,
+      spaces: this.spaces,
+      getServices: this.getServicesFactory(core.savedObjects),
+      encryptedSavedObjectsPlugin: plugins.encryptedSavedObjects,
+      actionTypeRegistry: actionTypeRegistry!,
+    });
+
+    taskRunnerFactory!.initialize({
+      encryptedSavedObjectsPlugin: plugins.encryptedSavedObjects,
+      getBasePath: this.getBasePath,
+      spaceIdToNamespace: this.spaceIdToNamespace,
+    });
+
+    return {
+      execute: createExecuteFunction({
+        taskManager: plugins.taskManager,
+        getScopedSavedObjectsClient: core.savedObjects.getScopedClient,
+        getBasePath: this.getBasePath,
+      }),
+    };
+  }
+
+  private getServicesFactory(
+    savedObjects: SavedObjectsServiceStart
+  ): (request: KibanaRequest) => Services {
+    const { adminClient } = this;
+    return request => ({
+      callCluster: adminClient!.asScoped(request).callAsCurrentUser,
```

Comment:
My fear is that we're going to end up with dozens of these, and it's annoying because this would not be an issue if it wasn't for the use of a class.
Fields are always nullable if you don't set them at constructor level, but we can't do that because of the NP plugin API.

I'm thinking of forcing this via schema validation of all plugins, but that should probably be a separate PR.
Any thoughts?

Reply:
huh, yeah, would be an interesting approach.  We'd have some bag of all the potential nullables, pass them into a validator which returns the non-null versions?  That might also get unwieldy, but seems like it's worth a try.